gui/citylist_frame_t.cc

```diff
@@ -244,9 +255,41 @@ bool citylist_frame_t::action_triggered( gui_action_creator_t *comp, value_t v)
 	else if( comp == &filter_by_owner ) {
 		fill_list();
 	}
+	else if( comp == &bt_copy_csv ) {
+		copy_csv_format();
+	}
 	return true;
 }
 
+// copy citylist and respective populations to clipboard
+void citylist_frame_t::copy_csv_format() {
+	// copy in csv format. separator is \t.
+	cbuffer_t clipboard;
+	
+	// append header
+	clipboard.append("City\tpopulation\n");
+	
+	// loop over the cities shown in the city list and add the name/population size
+	for (int i = 0; i < scrolly.get_count(); i++) {
+
+		// get city from city list
+		const citylist_stats_t* element = dynamic_cast<const citylist_stats_t*>(scrolly.get_element(i));
+		stadt_t* city = element->get_city();
+
+		// add to clipboard
+		clipboard.printf( "%s\t", city->get_name() );
+		clipboard.printf( "%d\n", city->get_einwohner());
+	}
+
+	// tell user if successfull or not
+	if(  clipboard.len()>0  ) {
+		dr_copy( clipboard, clipboard.len() );
+		create_win( new news_img("Population data was copied to clipboard.\n"), w_time_delete, magic_none );
+	} else {
+		create_win( new news_img("There is nothing to copy.\n"), w_time_delete, magic_none );
+	}
```

Comment:
Since always the text `"City\tpopulation\n"` is appended to the clipboard, will `clipboard.len()` be always larger than zero?

Reply:
I have always learned that it is good practice to always expect the unexpected and make exception cases for if things like lengths are not what you are expecting. After all if this wasn't there and the length somehow had become 0 or negative (due to some error in different code or something else like a corruption of the clipboard or whatever) the game will crash.

Reply:
> I have always learned that it is good practice to always expect the unexpected

This is true for external inputs. However in this time, the assumption `clipboard.len()>0` is statically derived from the code rather than some assumptions about the external inputs.

I think it's still reasonable to check `clipboard.len()>0` to prevent a crash, but please leave a comment to indicate that the check is for just-in-case. If you want to make the programming error detectable easier, please use `dbg->error` instead of showing `There is nothing to copy` message to users.